admin-dev/themes/new-theme/js/components/grid/extension/async-submit-bulk-action-extension.js

```diff
@@ -0,0 +1,238 @@
+      extend: (grid) => this.extend(grid),
+    };
+  }
+
+  setSuccessCallback(callback) {
+    return this.isCallback(callback) ? callback
+      : (response) => {
+        if ('message' in response) {
+          window.showSuccessMessage(response.message);
+        }
+
+        window.location.reload();
+      };
+  }
+
+  setErrorCallback(callback) {
+    return this.isCallback(callback) ? callback
+      : (response) => {
+        if ('message' in response) {
+          window.showErrorMessage(response.message);
+        }
+
+        window.location.reload();
+      };
+  }
+
+  isCallback(callback) {
+    return callback && typeof callback === 'function';
+  }
+
+  /**
+   * Extend grid with bulk action submitting
+   *
+   * @param {Grid} grid
+   */
+  extend(grid) {
+    grid.getContainer().on('click', '.js-bulk-action-submit-async-btn', (event) => {
+      this.submit(event, grid);
+    });
+  }
+
+  /**
+   * Collects required data and submits the form.
+   * @param event
+   * @param grid
+   */
+  submit(event, grid) {
+    const $submitBtn = $(event.currentTarget);
+    const confirmMessage = $submitBtn.data('confirm-message');
+    const confirmTitle = $submitBtn.data('confirmTitle');
+    const hasConfirmTitle = confirmTitle !== undefined;
+    const hasConfirmationMessage = confirmMessage !== undefined && confirmMessage.length > 0;
+    const hasAllConfirmationMessages = hasConfirmationMessage && hasConfirmTitle;
+
+    if (hasAllConfirmationMessages) {
+      this.showConfirmModal($submitBtn, grid, confirmMessage, confirmTitle);
+    } else if (hasConfirmationMessage && window.confirm(confirmMessage)) {
+      this.submitForm($submitBtn, grid);
+    } else {
+      this.submitForm($submitBtn, grid);
+    }
+  }
+
+  submitForm($submitBtn, grid) {
+    const $form = $(`#${grid.getId()}_filter_form`);
+    const $checkedInputs = this.getCheckedInputs($form);
+    const ids = this.getCheckedIds($checkedInputs);
+
+    const inputName = this.getInputName($checkedInputs);
+    const chunkSize = parseInt($submitBtn.data('chunkSize'), 10);
+    const chunkedIds = this.chunkArray([...ids], chunkSize);
+    const submitMethod = $submitBtn.data('form-method');
+    const formUrl = $submitBtn.data('form-url');
+
+    if (!['POST', 'GET', 'DELETE', 'PUT'].includes(submitMethod)) {
+      return;
+    }
+
+    /**
+     * Async submit action which repeats itself in recursive order.
+     * @param items
+     * @param successCallback
+     * @param errorCallback
+     */
+    const asyncSubmit = (items, successCallback, errorCallback) => {
+      if (items.length === 0) {
+        return;
+      }
+
+      const firstIdsChunk = items.shift();
+      const data = {};
+      data[inputName] = firstIdsChunk;
+
+      $.ajax({
+        type: submitMethod,
+        url: formUrl,
+        data,
+        dataType: 'json',
+      }).then((response) => {
+        this.progressCallback(response, firstIdsChunk, ids);
+        if (!response.success) {
```

Comment:
You should use an `else if` after this statement because in case the last call has an error, the condition `response.success && items.length > 0` won't be valid, then it will go in the `else` So you will consecutively call `errorCallback` and `successCallback` which may be confusing ^^

Besides as I told you in a previous comment, I don't think you should stop the loop in case of one mistake The bulk action may fail on only one out of ten products but you should try performing it for the remaining ones

So if you manage the success/error callbacks in the `then` and `catch` method, but systematically perform the `asyncSubmit` check and call in the `always` method, then you have proper display of error/success messages and you still try to loop through all the product list